File: app/src/main/java/com/scatl/uestcbbs/module/credit/view/CreditHistoryActivity.java
package com.scatl.uestcbbs.module.credit.view;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.TextView;

import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.annotation.ToastType;
import com.scatl.uestcbbs.base.BaseActivity;
import com.scatl.uestcbbs.base.BasePresenter;
import com.scatl.uestcbbs.callback.OnRefresh;
import com.scatl.uestcbbs.custom.MyLinearLayoutManger;
import com.scatl.uestcbbs.entity.MineCreditBean;
import com.scatl.uestcbbs.module.credit.adapter.MineCreditHistoryAdapter;
import com.scatl.uestcbbs.module.credit.presenter.CreditHistoryPresenter;
import com.scatl.uestcbbs.module.credit.presenter.MineCreditPresenter;
import com.scatl.uestcbbs.module.dayquestion.view.DayQuestionActivity;
import com.scatl.uestcbbs.module.magic.view.MagicShopActivity;
import com.scatl.uestcbbs.module.medal.view.MedalCenterActivity;
import com.scatl.uestcbbs.util.Constant;
import com.scatl.uestcbbs.util.RefreshUtil;
import com.scatl.uestcbbs.util.TimeUtil;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.constant.RefreshState;

import java.util.List;

public class CreditHistoryActivity extends BaseActivity implements CreditHistoryView{

    Toolbar toolbar;
    CreditHistoryPresenter creditHistoryPresenter;
    MineCreditHistoryAdapter mineCreditHistoryAdapter;
    RecyclerView recyclerView;
    SmartRefreshLayout refreshLayout;

    int page = 1;

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_credit_history;
    }

    @Override
    protected void findView() {
        toolbar = findViewById(R.id.toolbar);
        recyclerView = findViewById(R.id.credit_history_rv);
        refreshLayout = findViewById(R.id.credit_history_refresh);
    }

    @Override
    protected void initView() {
        super.initView();
        creditHistoryPresenter = (CreditHistoryPresenter) presenter;

        mineCreditHistoryAdapter = new MineCreditHistoryAdapter(R.layout.item_credit_history);
        mineCreditHistoryAdapter.setHasStableIds(true);
        recyclerView.setLayoutManager(new MyLinearLayoutManger(this));
        recyclerView.setAdapter(mineCreditHistoryAdapter);
        recyclerView.setLayoutAnimation(AnimationUtils.loadLayoutAnimation(this, R.anim.layout_animation_scale_in));

        refreshLayout.autoRefresh(0, 300, 1, false);
    }

    @Override
    protected BasePresenter initPresenter() {
        return new CreditHistoryPresenter();
    }


    @Override
    protected void setOnRefreshListener() {
        RefreshUtil.setOnRefreshListener(this, refreshLayout, new OnRefresh() {
            @Override
            public void onRefresh(RefreshLayout refreshLayout) {
                page = 1;
                creditHistoryPresenter.getCreditHistory(page);
            }

            @Override
            public void onLoadMore(RefreshLayout refreshLayout) {
                creditHistoryPresenter.getCreditHistory(page);
            }
        });
    }


    @Override
    public void onGetMineCreditHistorySuccess(List<MineCreditBean.CreditHistoryBean> creditHistoryBeans, boolean hasNext) {
        if (page == 1) {
            mineCreditHistoryAdapter.setNewData(creditHistoryBeans);
            recyclerView.scheduleLayoutAnimation();
        } else {
            mineCreditHistoryAdapter.addData(creditHistoryBeans);
        }


        if (refreshLayout.getState() == RefreshState.Refreshing) {
            page = 2;
            if (hasNext) {
                refreshLayout.finishRefresh();
            } else {
                refreshLayout.finishRefreshWithNoMoreData();
            }
        }
        if (refreshLayout.getState() == RefreshState.Loading) {
            if (hasNext) {
                page = page + 1;
                refreshLayout.finishLoadMore(true);
            } else {
                refreshLayout.finishLoadMoreWithNoMoreData();
            }
        }
    }

    @Override
    public void onGetMineCreditHistoryError(String msg) {
        refreshLayout.finishLoadMore();
        refreshLayout.finishRefresh();
        showToast(msg, ToastType.TYPE_ERROR);
    }
}

File: app/src/main/java/com/scatl/uestcbbs/module/credit/view/CreditHistoryActivity.kt
package com.scatl.uestcbbs.module.credit.view

import android.graphics.Rect
import android.view.HapticFeedbackConstants
import android.view.View
import android.view.animation.AnimationUtils
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.base.BaseVBActivity
import com.scatl.uestcbbs.databinding.ActivityCreditHistoryBinding
import com.scatl.uestcbbs.entity.MineCreditBean
import com.scatl.uestcbbs.module.credit.adapter.CreditHistoryAdapter
import com.scatl.uestcbbs.module.credit.presenter.CreditHistoryPresenter
import com.scatl.uestcbbs.util.CommonUtil
import com.scatl.uestcbbs.widget.span.CustomClickableSpan
import com.scatl.util.ScreenUtil
import com.scwang.smart.refresh.layout.api.RefreshLayout

class CreditHistoryActivity : BaseVBActivity<CreditHistoryPresenter, CreditHistoryView, ActivityCreditHistoryBinding>(), CreditHistoryView {

    private lateinit var creditHistoryAdapter: CreditHistoryAdapter
    private var mPage = 1
    private var mCurrentInOutSort = 0
    private var mCurrentCreditType = 0

    override fun getViewBinding() = ActivityCreditHistoryBinding.inflate(layoutInflater)

    override fun initPresenter() = CreditHistoryPresenter()

    override fun initView(theftProof: Boolean) {
        super.initView(false)
        creditHistoryAdapter = CreditHistoryAdapter(R.layout.item_credit_history)
        mBinding.recyclerView.apply {
            adapter = creditHistoryAdapter
            layoutAnimation = AnimationUtils.loadLayoutAnimation(this@CreditHistoryActivity, R.anim.layout_animation_scale_in)
            addItemDecoration(object : ItemDecoration() {
                override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
                    outRect.bottom = ScreenUtil.dip2px(context, 10f)
                }
            })
        }

        mBinding.inOutGroup.check(R.id.default_in_out_sort_btn)
        mBinding.creditTypeGroup.check(R.id.default_credit_sort_btn)
        bindClickEvent(
            mBinding.defaultInOutSortBtn, mBinding.inSortBtn, mBinding.outSortBtn,
            mBinding.defaultCreditSortBtn, mBinding.waterSortBtn, mBinding.weiwangSortBtn, mBinding.jiangliquanSortBtn
        )

        mBinding.statusView.success()

        mBinding.refreshLayout.autoRefresh(0, 300, 1f, false)
    }

    override fun onClick(v: View) {
        when(v) {
            mBinding.defaultInOutSortBtn -> { mCurrentInOutSort = 0 }
            mBinding.inSortBtn -> { mCurrentInOutSort = 1 }
            mBinding.outSortBtn -> { mCurrentInOutSort = -1 }

            mBinding.defaultCreditSortBtn -> { mCurrentCreditType = 0 }
            mBinding.weiwangSortBtn -> { mCurrentCreditType = 1 }
            mBinding.waterSortBtn -> { mCurrentCreditType = 2 }
            mBinding.jiangliquanSortBtn -> { mCurrentCreditType = 6 }
        }
        mPage = 1
        v.performHapticFeedback(HapticFeedbackConstants.CONTEXT_CLICK)
        mBinding.refreshLayout.autoRefresh(0, 300, 1f, false)
        mBinding.recyclerView.layoutManager?.scrollToPosition(0)
    }

    override fun setOnItemClickListener() {
        creditHistoryAdapter.setOnItemClickListener { adapter, view, position ->
            creditHistoryAdapter.data[position].link?.let {
                CustomClickableSpan(getContext(), it).onClick(View(getContext()))
            }
        }
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        mPage = 1
        mPresenter?.getCreditHistory(mPage, mCurrentCreditType, mCurrentInOutSort)
    }

    override fun onLoadMore(refreshLayout: RefreshLayout) {
        mPresenter?.getCreditHistory(mPage, mCurrentCreditType, mCurrentInOutSort)
    }

    override fun onGetMineCreditHistorySuccess(creditHistoryBeans: List<MineCreditBean.CreditHistoryBean>, hasNext: Boolean) {
        mBinding.statusView.success()
        mBinding.refreshLayout.finishRefresh()

        if (mPage == 1) {
            creditHistoryAdapter.setNewData(creditHistoryBeans)
            mBinding.recyclerView.scheduleLayoutAnimation()
        } else {
            creditHistoryAdapter.addData(creditHistoryBeans)
        }

        if (hasNext) {
            mPage ++
            mBinding.refreshLayout.finishLoadMore(true)
        } else {
            mBinding.refreshLayout.finishLoadMoreWithNoMoreData()
        }
    }

    override fun onGetMineCreditHistoryError(msg: String?) {
        mBinding.refreshLayout.finishRefresh()
        mBinding.refreshLayout.finishLoadMore(false)
        creditHistoryAdapter.setNewData(mutableListOf())
        if (mPage == 1) {
            mBinding.statusView.error(msg)
        }
    }

    override fun getContext() = this

}